### PR TITLE
Desktop: Fix ENEX → HTML import

### PR DESCRIPTION
**Pull request by devonzuegel**
### Problem
I was importing a notebook yesterday and discovered that `import-enex.js` hangs when importing a massive resource (e.g. a video or big PDF) attached to an Evernote note. 

The bug appeared as a lack of a `created_time` and `updated_time` on the note object, but after a fair bit of debugging it turns out that the real problem lays in the moment when the `noteResource.data` is loaded into the buffer:

```
decodedData = Buffer.from(noteResource.data, 'base64');
```

This simply happened to manifest in messing up the parsing for the next note. For example, if you have three notes that we'll call A, B, and C, and B has a big video resource attached, then 
(a) B will fail to attach the video, 
(b) C will fail to parse because the importer thinks it doesn't have a `created_time` and `updated_time` (even though it does!), and 
(c) the app will hang and the screen will turn white.

You can reproduce this bug by following these steps:
1. Create an Evernote note with a big video file (141 MB did it for me). image
2. Export that note with one before it and one after.
3. Import the `.enex` file into Joplin. It should hang at this step and in the console say that it was disconnected.

### Description of changes
Primary change:
- The primary change in this PR is that it does not attempt to load data into a buffer if the data is massive. It just straight up gives up on those. In those cases, it logs a warning to the console and leaves an inline warning that looks like this:

  image

  **Impact:** When I tested it on one of my real-world notebooks that has 960 notes, only 3 of them were impacted. It was trivial enough to go back, look at the warnings, and simply manually drag-and-drop the missed attachmnents from Evernote into the Joplin notes. This is an extremely hacky workaround, but it's better than the current state of things.

The PR introduces a few other small changes too:
- It logs import progress to the console for ENEX to HTML imports: image
- It adds a hotkey accelerator to the ENEX → HTML import. This was very helpful for tightening my feedback loop as I fixed this bug. Since imports are generally a fairly infrequent task, it's probably not so important that we have a hotkey here, but it was available so I figured I might as well leave it. Happy to remove it if you prefer though.
- It adds a bright red `deleteAllFoldersButton` at the bottom of the sidebar for easy cleanup of notebooks during development. @laurent22 is there a way to hide this under a development flag so it only shows when in debug mode or something? If not, we can just remove it. image

### Notes
I thought there was some chance that there was another issue with the specific video format (`.m4v`), so to check to make sure that it's not because it's an `.m4v` format and that it actually was the size issue, I tested the import on master with [this sample](https://github.com/robovm/apple-ios-samples/blob/master/AVTimedAnnotationWriterUsingCustomAnnotationMetadataforMovieWritingandPlayback/AVTimedAnnotationWriter/sample.m4v), and it worked fine. So it appears it really is a size issue. (Or at least it's not an issue with importing `.m4v`s.)

### Next steps
There's surely a way to load the massive resources properly, but I didn't have time to go down that path and figured the partial fix here was at least a step in the right direction.

### CODE REVIEW

**Comment by devonzuegel** [comment]
There's still one test failure, but I'm not quite sure how it relates to the changes I made: 

image

File: ReactNativeClient/PluginAssetsLoader.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+		step((generator = generator.apply(thisArg, _arguments || [])).next());
+	});
+};
+Object.defineProperty(exports, '__esModule', { value: true });
+const { dirname } = require('lib/path-utils.js');
+const { shim } = require('lib/shim');
+const Setting = require('lib/models/Setting');
+const pluginAssets = require('./pluginAssets/index');
+const KvStore = require('lib/services/KvStore.js');
+class PluginAssetsLoader {
```
Comment (laurent22) [inline ReactNativeClient/PluginAssetsLoader.js]: This file shouldn't be committed (I think if you rebase, it's now included in .gitignore).

File: ReactNativeClient/lib/import-enex.js [hunk 2]
```diff
@@ -306,7 +311,7 @@ function importEnex(parentFolderId, filePath, importOptions = null) {
 			if (n == 'note') {
 				note = {
 					resources: [],
-					tags: [],
+					tags: ['enex import', `enex to ${importOptions.outputFormat} import`],
```
Comment (laurent22) [inline ReactNativeClient/lib/import-enex.js]: One thing I learnt with the Welcome notes fiasco, is that users don't like auto-created tags so I would prefer if we don't create these.

**Comment by laurent22** [comment]
Hmm, with this change it's indeed better than it was before, but I would prefer to see first if there's a way to import the resources, regardless of their size. I'm guessing we could read the base64 data chunk by chunk and write it to a temp file (as opposed to memory, like now), and then when the resource is created we copy this file directly. I might give it a try.

There are other more straightforward changes that we could merge though, like the progress report and delete all button. By the way, you can use `Setting.value('env')` ("dev" or "prod") to check the environment.

**Comment by laurent22** [comment]
@devonzuegel, would you mind uploading a sample note with a large attachment somewhere, so that I can try to replicate? With my Evernote account I can only attach up to 25MB to note and I assume it wouldn't be enough to crash the app.

**Comment by laurent22** [comment]
@devonzuegel, I have fixed the issue with importing large attachments from ENEX file in this commit: 9ec1e84ed09cd362f1a9e786473adae5e33e0486 The issue was partly a bug in the Sax package, which had a memory leak, and partly a bug in Joplin's code, which also had a memory leak. I've fixed it by patching Sax and using streams to decode the resources, so memory usage should be stable now matter how big the resource is. I've tried with a 1.5GB file with no issue.

What would you like to do about this PR, should we close it or would you like to get the other changes in? The log update and Delete All button would still be useful, but it can also be as part of a separate PR, just let us know.

**Comment by laurent22** [comment]
@devonzuegel, as we're getting a lot of activity due to GSoC at the moment, I'm trying to sort out the older PR so if you don't mind I'll close this one. We can re-open if you'd like to keep working on it of course, although I feel it would be best to take out the features you'd like to keep to a new branch and create a PR from that. Anyway, please let us know if you'd like to do something about it.

**Pull request closed without merging**